Common/include/toolboxes/graph_toolbox.hpp

```diff
@@ -0,0 +1,489 @@
+    while(m_innerIdx(k) != iInnerIdx) ++k;
+    return k;
+  }
+
+  /*!
+   * \param[in] iDiagIdx - Diagonal index (row == col).
+   * \return Absolute position of the diagonal entry.
+   */
+  inline Index_t getDiagPtr(Index_t iDiagIdx) const {
+    return m_diagPtr(iDiagIdx);
+  }
+
+  /*!
+   * \return Raw pointer to the outer pointer vector.
+   */
+  inline const Index_t* outerPtr() const {
+    assert(!empty() && "Sparse pattern has not been built.");
+    return m_outerPtr.data();
+  }
+
+  /*!
+   * \return Raw pointer to the inner index vector.
+   */
+  inline const Index_t* innerIdx() const {
+    assert(!empty() && "Sparse pattern has not been built.");
+    return m_innerIdx.data();
+  }
+
+  /*!
+   * \return Raw pointer to the diagonal pointer vector.
+   */
+  inline const Index_t* diagPtr() const {
+    assert(!m_diagPtr.empty() && "Diagonal map has not been built.");
+    return m_diagPtr.data();
+  }
+
+  /*!
+   * \return The minimum inner index.
+   */
+  Index_t getMinInnerIdx() const {
+    Index_t idx = std::numeric_limits<Index_t>::max();
+    for(Index_t k=0; k<m_innerIdx.size(); ++k)
+      idx = std::min(idx, m_innerIdx(k));
+    return idx;
+  }
+
+  /*!
+   * \return The maximum inner index.
+   */
+  Index_t getMaxInnerIdx() const {
+    Index_t idx = std::numeric_limits<Index_t>::min();
+    for(Index_t k=0; k<m_innerIdx.size(); ++k)
+      idx = std::max(idx, m_innerIdx(k));
+    return idx;
+  }
+};
+
+/*!
+ * \brief Alias a type of container as the edge map class, this is a N by 2 container
+ *        that maps the two sparse entries referenced by an edge (ij and ji) to two
+ *        non zero entries of a sparse pattern.
+ */
+template<typename Index_t>
+using CEdgeToNonZeroMap = C2DContainer<unsigned long, Index_t, StorageType::RowMajor, 64, DynamicSize, 2>;
+
+
+using CCompressedSparsePatternUL = CCompressedSparsePattern<unsigned long>;
+using CEdgeToNonZeroMapUL = CEdgeToNonZeroMap<unsigned long>;
+
+
+/*!
+ * \brief Build a sparse pattern from geometry information, of type FVM or FEM,
+ *        for a given fill-level. At fill-level N, the immediate neighbors of the
+ *        points in level N-1 are also considered neighbors of the base point.
+ *        The resulting pattern is that of A^{N+1} where A is the sparse matrix
+ *        of immediate neighbors.
+ * \note Algorithm is equivalent to the implementation by F. Palacios,
+ *       A. Bueno, and T. Economon from CSysMatrix.
+ * \param[in] geometry - Definition of the grid.
+ * \param[in] type - Of connectivity.
+ * \param[in] fillLvl - Target degree of neighborhood (immediate neighbors always added).
+ * \return Compressed-Storage-Row sparse pattern.
+ */
+template<class Geometry_t, typename Index_t>
+CCompressedSparsePattern<Index_t> buildCSRPattern(Geometry_t& geometry,
+                                                  ConnectivityType type,
+                                                  Index_t fillLvl)
+{
```

Comment:
For kind of a change of pace I am building the sparse patterns in a stand-alone function instead of adding yet another method to CGeometry (the return by value is not an issue for that class in c++ 11).